core/src/Kokkos_LogicalMemorySpace.hpp

```diff
@@ -0,0 +1,358 @@
+                                const std::string& arg_label,
+                                const size_t arg_alloc_size) {
+    if (!arg_alloc_size) return (void*)0;
+
+    SharedAllocationRecord* const r =
+        allocate(arg_space, arg_label, arg_alloc_size);
+
+    RecordBase::increment(r);
+
+    return r->data();
+  }
+
+  /**\brief  Reallocate tracked memory in the space */
+  static void* reallocate_tracked(void* const arg_alloc_ptr,
+                                  const size_t arg_alloc_size) {
+    SharedAllocationRecord* const r_old = get_record(arg_alloc_ptr);
+    SharedAllocationRecord* const r_new =
+        allocate(r_old->m_space, r_old->get_label(), arg_alloc_size);
+
+    Kokkos::Impl::DeepCopy<SpaceType, SpaceType>(
+        r_new->data(), r_old->data(), std::min(r_old->size(), r_new->size()));
+
+    RecordBase::increment(r_new);
+    RecordBase::decrement(r_old);
+
+    return r_new->data();
+  }
+  /**\brief  Deallocate tracked memory in the space */
+  static void deallocate_tracked(void* const arg_alloc_ptr) {
+    if (arg_alloc_ptr != 0) {
+      SharedAllocationRecord* const r = get_record(arg_alloc_ptr);
+
+      RecordBase::decrement(r);
+    }
+  }
+
+  static SharedAllocationRecord* get_record(void* alloc_ptr) {
+    typedef SharedAllocationHeader Header;
+    typedef SharedAllocationRecord<SpaceType, void> RecordHost;
+
+    SharedAllocationHeader const* const head =
+        alloc_ptr ? Header::get_header(alloc_ptr) : (SharedAllocationHeader*)0;
+    RecordHost* const record =
+        head ? static_cast<RecordHost*>(head->m_record) : (RecordHost*)0;
+
+    if (!alloc_ptr || record->m_alloc_ptr != head) {
+      Kokkos::Impl::throw_runtime_exception(
+          std::string("Kokkos::Impl::SharedAllocationRecord< SpaceType , "
+                      "void >::get_record ERROR"));
+    }
+
+    return record;
+  }
+#ifdef KOKKOS_DEBUG
+  static void print_records(std::ostream& s, const SpaceType&,
+                            bool detail = false) {
+    SharedAllocationRecord<void, void>::print_host_accessible_records(
+        s, "HostSpace", &s_root_record, detail);
+  }
+#else
+  static void print_records(std::ostream&, const SpaceType&,
+                            bool detail = false) {
+    throw_runtime_exception(
+        "SharedAllocationRecord<HostSpace>::print_records only works with "
+        "KOKKOS_DEBUG enabled");
+  }
+#endif
+};
+#ifdef KOKKOS_DEBUG
+/**\brief  Root record for tracked allocations from this HostSpace instance */
+template <const char* Name, class BaseSpace, class DefaultExecutionSpace,
+          bool SharesAccessSemanticsWithBase>
+RecordBase<BaseSpace, DefaultExecutionSpace,
+           SharesAccessSemanticsWithBase>::s_root_record;
+#endif
+
+}  // namespace Impl
+
+}  // namespace Kokkos
+
+//----------------------------------------------------------------------------
+
+namespace Kokkos {
+
+namespace Impl {
+
+#define PAR_DEEP_COPY_USE_MEMCPY
```

Comment:
Why is this defined here? Even if it's supposed to be defined, shouldn't it be `#undef`'d somewhere else before leaving this file?  It doesn't look like it does anything to this file if you define it here...

Reply:
Also pretty sure this is copy-pasta'd